### ISSUE TITLE
Why pass the scores through a softmax layer in EnsembleLogitsProcessor?

### ISSUE THREAD

**Issue by liuhaozhe6788** [ZurichNLP/ContraDecode#1]
Hello, thank you for the great work on this project.
I have a question regarding the implementation of EnsembleLogitsProcessor in translation_models/m2m100.py. I noticed that within the processor, the scores are passed through a softmax operation. However, based on my understanding of the beam search implementation in the Transformers library, the scores at this stage are already log-probabilities rather than raw logits.
Could you please clarify the rationale for applying softmax in this context? I would appreciate any insight you can provide. Thank you in advance!

**Comment by jvamvas** [comment]
Hi @liuhaozhe6788, thanks for raising this issue!

Formula 1 in the paper defines that the subtraction is performed in prob-space:

<img width="300" alt="Image" src="https://github.com/user-attachments/assets/24d807d6-8020-4ae4-9670-e4d90b290d68" />

This requires us to apply softmax before we perform the subtraction, and to convert back to log-space after that (line https://github.com/ZurichNLP/ContraDecode/blob/main/translation_models/m2m100.py#L44).

Hope this helps!

**Comment by liuhaozhe6788** [comment]
Hello, @jvamvas. Yes, I see. But I checked the implementation of beam search in [https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L3939](url), and before the scores are passed into a logit processor, they are already log-probabilities.
